File: java/ApkResourceParser/apk-resource-parser/src/main/java/ikas/project/java/apk/arsc/Preconditions.java
package ikas.project.java.apk.arsc;

public class Preconditions {
    public static void checkArgument(boolean expression, /*@CheckForNull*/ Object errorMessage) {
        if (!expression) {
            var message = switch (errorMessage) {
                case null -> "";
                case String s -> s;
                case Object o -> o.toString();
            };
            throw new IllegalArgumentException(message);
        }
    }

    public static void checkState(boolean expression) {
        if (!expression) {
            throw new IllegalStateException();
        }
    }

    public static void checkState(boolean expression, /*@CheckForNull*/ Object errorMessage) {
        if (!expression) {
            var message = switch (errorMessage) {
                case null -> "";
                case String s -> s;
                case Object o -> o.toString();
            };
            throw new IllegalStateException(message);
        }
    }

    public static <T> T checkNotNull(/*@CheckForNull*/ T reference) {
        if (reference == null) {
            throw new NullPointerException();
        }
        return reference;
    }

    public static <T> T checkNotNull(/*@CheckForNull*/ T reference, /*@CheckForNull*/ Object errorMessage) {
        if (reference == null) {
            var message = switch (errorMessage) {
                case null -> "";
                case String s -> s;
                case Object o -> o.toString();
            };
            throw new NullPointerException(message);
        }
        return reference;
    }

    public static <T> T checkNotNull(
            /*@CheckForNull*/ T reference,
                              String errorMessageTemplate,
            /*@CheckForNull*/ /*@Nullable*/ Object... errorMessageArgs) {
        if (reference == null) {
            throw new NullPointerException(String.format(errorMessageTemplate, errorMessageArgs));
        }
        return reference;
    }

}

File: dotnet/ApkResouceParser.NetStandard2/Preconditions.cs
/*
 *	apk resource parser for .net
 *  port of https://github.com/google/android-arscblamer
 *
 *  ikas-mc 2023
 */

using System;

namespace ApkResourceParser
{
    public static class Preconditions
    {
        public static void checkArgument(bool expression, object errorMessage)
        {
            if (!expression)
            {
                string message;
                switch (errorMessage)
                {
                    case null:
                        message = "";
                        break;
                    case string s:
                        message = s;
                        break;
                    default:
                        message = errorMessage.ToString();
                        break;
                }
                throw new ArgumentException(message);
            }
        }

        public static void checkState(bool expression)
        {
            if (!expression)
            {
                throw new ArgumentException();
            }
        }

        public static void checkState(bool expression, object errorMessage)
        {
            if (!expression)
            {
                string message;
                switch (errorMessage)
                {
                    case null:
                        message = "expression is false";
                        break;
                    case string s:
                        message = s;
                        break;
                    default:
                        message = errorMessage.ToString();
                        break;
                }
                throw new ArgumentException(message);
            }
        }

        public static T checkNotNull<T>(T reference)
        {
            if (reference == null)
            {
                throw new ArgumentException();
            }

            return reference;
        }

        public static T checkNotNull<T>(T reference, object errorMessage)
        {
            if (reference == null)
            {
                string message;
                switch (errorMessage)
                {
                    case null:
                        message = "value is null";
                        break;
                    case string s:
                        message = s;
                        break;
                    default:
                        message = errorMessage.ToString();
                        break;
                }
                throw new ArgumentException(message);
            }

            return reference;
        }
    }
}